### PR TITLE
Fix: Removed false positives of break and continue (fixes #5972)

### LINKED ISSUE #5972: 'no-unsafe-finally' rule gives errors on safe code
**What version of ESLint are you using?**
eslint@e606523

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**What did you do? Please include the actual source code causing the issue.**
Rule `no-unsafe-finally` gives an error on the following code:
```js
try {} finally {
  while(true) {
    break;
  }
}
```

**What did you expect to happen?**
No errors.

**What actually happened? Please include the actual, raw output from ESLint.**
`Unsafe usage of BreakStatement                          no-unsafe-finally`

More generally, break and continue statements need to check that the statement being broken or continued is not interior to the finally block. This involves checking labels, if present.

ping @onurtemizkan, @platinumazure.

See #5808, #5932.

### PR DESCRIPTION
**Pull request by onurtemizkan**
fixes: #5972

### CODE REVIEW

**Comment by mysticatea** [comment]
Thank you for this PR!

I want to add more tests.

- valid:
  - `try {} finally { while (true) break; }`
  - `try {} finally { while (true) continue; }`
  - `try {} finally { switch (true) { case true: break; } }`
- invalid
  - `while (true) try {} finally { break; }`
  - `while (true) try {} finally { continue; }`
  - `switch (true) { case true: try {} finally { break; } }`

File: lib/rules/no-unsafe-finally.js [hunk 1]
```diff
@@ -9,7 +9,8 @@
 // Helpers
 //------------------------------------------------------------------------------
 
-var SENTINEL_NODE_TYPE = /^(?:Program|(?:Function|Class)(?:Declaration|Expression)|ArrowFunctionExpression)$/;
+var SENTINEL_NODE_TYPE_RETURN_THROW = /^(?:Program|(?:Function|Class)(?:Declaration|Expression)|ArrowFunctionExpression)$/;
+var SENTINEL_NODE_TYPE_BREAK_CONTINUE = /^(?:Program|(?:Function|Class)(?:Declaration|Expression)|ArrowFunctionExpression|WhileStatement|ForStatement|SwitchStatement)$/;
```
Comment (onurtemizkan) [inline lib/rules/no-unsafe-finally.js]: Oops, forgotten do while statements. :laughing:

File: lib/rules/no-unsafe-finally.js [hunk 2]
```diff
@@ -9,7 +9,8 @@
 // Helpers
 //------------------------------------------------------------------------------
 
-var SENTINEL_NODE_TYPE = /^(?:Program|(?:Function|Class)(?:Declaration|Expression)|ArrowFunctionExpression)$/;
+var SENTINEL_NODE_TYPE_RETURN_THROW = /^(?:Program|(?:Function|Class)(?:Declaration|Expression)|ArrowFunctionExpression)$/;
```
Comment (BYK) [inline lib/rules/no-unsafe-finally.js]: I think you can omit the outer most parenthesis here.

File: lib/rules/no-unsafe-finally.js [hunk 3]
```diff
@@ -9,7 +9,8 @@
...
+var SENTINEL_NODE_TYPE_BREAK_CONTINUE = /^(?:Program|(?:Function|Class)(?:Declaration|Expression)|ArrowFunctionExpression|DoWhileStatement|WhileStatement|ForStatement|SwitchStatement)$/;
```
Comment (BYK) [inline lib/rules/no-unsafe-finally.js]: Same as above, I don't think you need the outer most parenthesis.

Comment (onurtemizkan) [inline lib/rules/no-unsafe-finally.js]: Updated both :blush:

Comment (onurtemizkan) [inline lib/rules/no-unsafe-finally.js]: @byk Noticed I should also add ForOfStatement, ForInStatement etc. The regex will get very messy. Should I look for another solution?

File: lib/rules/no-unsafe-finally.js [hunk 4]
```diff
@@ -9,7 +9,8 @@
 // Helpers
 //------------------------------------------------------------------------------
 
-var SENTINEL_NODE_TYPE = /^(?:Program|(?:Function|Class)(?:Declaration|Expression)|ArrowFunctionExpression)$/;
+var SENTINEL_NODE_TYPE_RETURN_THROW = /^(?:Program)|(?:Function|Class)(?:Declaration|Expression)|ArrowFunctionExpression$/;
+var SENTINEL_NODE_TYPE_BREAK_CONTINUE = /^(?:Program)|(?:Function|Class)(?:Declaration|Expression)|ArrowFunctionExpression|DoWhileStatement|WhileStatement|ForStatement|SwitchStatement$/;
```
Comment (mysticatea) [inline lib/rules/no-unsafe-finally.js]: ForInStatement and ForOfStatement.

**Comment by mysticatea** [comment]
Thank you for the update!

I'd like to see more tests:

- invalid:
  - `a: while (true) try {} finally { while (true) break a; }`
  - `a: while (true) try {} finally { while (true) continue a; }`
  - `a: switch (true) { case true: try {} finally { switch (true) { case true: break a; } } }`
  - `a: while (true) try {} finally { switch (true) { case true: break a; } }`
  - `a: while (true) try {} finally { switch (true) { case true: continue; } }`

File: lib/rules/no-unsafe-finally.js (hunk 4 above)
Comment (mysticatea) [inline lib/rules/no-unsafe-finally.js]: Ah, and the parenthesis `/^(?: .... )$/` is needed.

`/^aaa|bbb$/` is meaning `^aaa` or `bbb$`.

Comment (BYK) [inline lib/rules/no-unsafe-finally.js]: @mysticatea oh, dang. You're right. May be we need a rule to warn on the "invalid" one then? :P

File: lib/rules/no-unsafe-finally.js (hunk 2 above)
Comment (platinumazure) [inline lib/rules/no-unsafe-finally.js]: Not so- the anchors would bind only to the first and last alternations in that case.

**Comment by mysticatea** [comment]
LGTM.

Thank you for great works!

File: lib/rules/no-unsafe-finally.js (hunk 4 above)
Comment (mysticatea) [inline lib/rules/no-unsafe-finally.js]: Indeed, it's one of general mistakes.
I will make the proposal later.

**Pull request merged**